proposals/20230316-oci-content-source.md

```diff
@@ -0,0 +1,284 @@
+
+#### Artifact Types
+Image manifests will be differentiated based on the `artifactType` field in the image manifest. The following media types will be used.
+- Component: "application/vnd.wasm.component.v1"
+- Interface: "application/vnd.wasm.component.interface.v1"
+- Bundled Component: "application/vnd.wasm.component.bundled.v1"
+
+#### Image Manifest for Components
+The following is an example image manifest for a component containing a configuration structure, a layer containing the `my-component.wasm` binary.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:5587da2246a78f08c447bff2ac91ee5c2b57be2f2a15244b5e618ac0be626885",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:b36aa5d0111a488937361fdb35432510d50675a11d566c5d9e82a147fb9ff552",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-component"
+      }
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
+  "architecture": "wasm32",
+  "os": "wasi"
+}
+```
+
+#### Image Manifest for Interface Components
+The following is an example image manifest for a component containing a configuration structure and a layer containing the `my-component-interface.wasm` binary.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.interface.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:c71d239df91726fc519c6eb72d318ec65820627232b2f796219e87dcf35d0ab4",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:dcf07c6bb395d6e1d40505b77e70af04f5fae0d54c9573fd379c1e7355a18cf3",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-component-interface"
+      }
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
+  "architecture": "wasm32",
+  "os": "wasi"
+}
+```
+
+#### Image Manifest for Bundled Components
+The following is an example image manifest for a component containing a configuration structure, a layer containing the `my-bundled-component.wasm` binary, and a data layer containing some a static asset.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.bundled.v1",
```

Comment:
Any particular reason for taking a hard dependency on OCI v1.1?
The alternative would be to only rely on the `config.mediaType` for now, and not mandate the top-level `artifactType`, which would make it possible to push to OCI v1.0-compatible registries.

Reply:
I was hoping someone would bring this up. I don't have a strong opinion. On one hand, I believe we should push for the most portable specification that fits our needs. On the other hand, I've received feedback that image manifests are not extremely specific and relying on the `config.mediaType` would be less expressive.

Reply:
Side note: looking at the manifest example again, it is not a valid manifest for either v1 or v1.1. There are two options:

- either `mediaType: application/vnd.oci.artifact.manifest.v1+json` and `artifactType: application/vnd.wasm.component.bundled.v1 `, and a list of `blobs` — OCI v1.1, not supported everywhere yet, but the better long term solution
OR
- setting `config.mediaType` and a list of `layers` (effectively ORAS) — which still has a top-level media type of a container image, but is accepted by _most_ registries.

Reply:
Perhaps, I'm misunderstanding the current work to remove `artifact.md` https://github.com/opencontainers/image-spec/pull/999, and the guidance for artifacts in https://github.com/opencontainers/image-spec/pull/1043, but it seems like artifact manifests will not be a thing in v1.1.

I tried to model this based on https://github.com/opencontainers/image-spec/pull/1043 section on [Guidelines for Artifact Usage](https://github.com/opencontainers/image-spec/blob/7ca9ebcaeffdf21ff6c5f274fbab611ee60578ef/manifest.md#guidelines-for-artifact-usage)

Reply:
Correct — if artifact manifests are not a thing in 1.1, then option 2 above is the only way to make this work for now (i.e. `artifactType` will not exist).

Reply:
n.b.: Setting `artifactType` is supported in OCI 1.0 and registries are out of conformance if they error on unknown fields. You can use both, and you can also set a scratch media type for `config.mediaType` if it's irrelevant to you. 

OCI 1.0 registries don't support the referrers API, so there shouldn't be any forwards or backwards compatibility issues.

Reply:
There's not a significant value add to having an `artifactType` when there is a dedicated `config.mediaType`. You'll find some registries have compatibility issues. And a lot of tooling hasn't been updated to read this yet.